박스오피스 [STEP1] Hisop

Hi-sop opened the pull request:
@havilog 
안녕하세요 STEP1 PR입니다.

## 고민이 되었던 점
중첩된 형태의 JSON이라 타입 구성이 조금 어려웠던 것 같습니다.

## 조언을 구하고 싶은 부분
- JSON의 key수가 많을 때 사용자의 오탈자에 의한 오류를 방지하는 팁이 있을까요?

- 만든 타입이 decoding을 정상적으로 진행할 수 있는지 테스트하고 싶었습니다
    동일한 값으로 JSON과 타입을 생성하고 decoding하여 생성한 타입과 비교하도록 했습니다.
    만국박람회 프로젝트때 사용했던 방식인데 이번 프로젝트에도 적절했을까요?

Hi-sop committed:
fork!

Hi-sop committed:
MVC패턴으로 폴더정리

Hi-sop committed:
feat: 타입 구현

약어x

Hi-sop committed:
feat: 테스트 추가

EqualTable extension 포함

havilog commented on BoxOffice/Model/BoxOfficeResult.swift:
```diff
@@ -0,0 +1,58 @@
+    let boxOfficeResult: BoxOfficeResult
+}
+
+struct BoxOfficeResult: Decodable {
+    let boxofficeType: String
+    let showRange: String
+    let dailyBoxOfficeList: [MovieStatistics]
+}
```
BoxOffice.Result와 같이 네스티드하게 쓰이는것도 고려해볼 수 있을거 같아요

Hi-sop replied:
네스티드하게 변경했습니다.

Hi-sop replied:
BoxOffice -> BoxOfficeResult -> DailyBoxOffice
차례대로 포함하고 있는 관계라 생각되어 모두 네스티드하게 변경했는데..! 변경하고 보니 읽기가 좀 어렵다는 느낌도 듭니다.
포함관계를 명확히 한다는 의미에서 이렇게 구성하는건가요? 추천해주신 이유가 궁금합니다

havilog commented on BoxOffice/Model/BoxOfficeResult.swift:
```diff
@@ -0,0 +1,58 @@
+struct BoxOfficeResult: Decodable {
+    let boxofficeType: String
+    let showRange: String
+    let dailyBoxOfficeList: [MovieStatistics]
```
프로퍼티 명은 데일리 박스오피스인데
타입은 무비 스태티스틱스인게 readablility가 떨어지는거 같습니다!
통일시켜주면 좋을거 같아요

Hi-sop replied:
DailyBoxOffice로 변경했습니다.

havilog commented on BoxOffice/Model/BoxOfficeResult.swift:
```diff
@@ -0,0 +1,58 @@
+    let screenCount: String
+    let showCount: String
+    
+    enum CodingKeys: String, CodingKey {
```
priavte 붙여주면 깨알 최적화

Hi-sop replied:
타입의 프로퍼티들에 private을 추가했습니다.

Hi-sop replied:
Equatable 프로토콜을 추가하는 extension에서도 접근이 불가능한 문제가 생겨 다시 제거해주었습니다.

havilog replied:
오잉.. 이 프로퍼티들에 붙이라는 이야기는 아니었구요 ..!
CodingKeys 에 붙이라는 얘기였어요!

Hi-sop replied:
추가완료했습니다!

havilog commented on BoxOffice/Model/BoxOfficeResult.swift:
```diff
@@ -0,0 +1,58 @@
+}
+
+struct MovieStatistics: Decodable {
+    let rankNumber: String
+    let rank: String
+    let rankIntensity: String
+    let rankOldAndNew: String
+    let movieCode: String
+    let movieName: String
+    let openDate: String
+    let salesAmount: String
+    let salesShare: String
+    let salesIntensity: String
+    let salesChange: String
+    let salesAccumulated: String
+    let audienceCount: String
+    let audienceIntensity: String
+    let audienceChange: String
+    let audienceAccumulated: String
+    let screenCount: String
+    let showCount: String
```
모든 프로퍼티들이 전부 UI를 그릴때 사용되는 아이들일까요?
서버에서만 사용하고 클라입장에서는 필요없는 프로퍼티일경우 굳이 선언하지 않아도 될 것 같습니다

Hi-sop replied:
현재는 테스트단계라고 판단하여 모두 넣어주었습니다. 일단 남겨두고 UI구현시에 필요한 프로퍼티를 제외하고 제거하겠습니다.

havilog commented on BoxOfficeTests/DecodingTests.swift:
```diff
@@ -0,0 +1,142 @@
+    }
+}
+
+extension BoxOffice: Equatable {
+    public static func == (lhs: BoxOffice, rhs: BoxOffice) -> Bool {
+        return lhs.boxOfficeResult == rhs.boxOfficeResult
+    }
+}
```
Equatability는 테스트쪽에서 익스텐션으로 해주는 것 보다는 본 타겟에서 해주는게 더 낫다고 생각해요

개인적인 생각으로는 대부분의 모델은 equatable해서 손해보는 부분은 많지 않고,
또한 Sendable과 Identifiable도 고려해서 모델 설계해주시면 좋을거 같습니다

Hi-sop replied:
Sendable
비동기 환경에서 특정 데이터를 보내도 안전한지 점검하는 프로토콜

Identifiable
프로토콜에 있는 프로퍼티 'id' 정의를 강제화

위 내용정도로 간단하게 이해했습니다. 

당장은 위 프로토콜들이 왜 쓰여야하는지 이해하지 못했습니다. 필요에 따라 추가하도록 설계해보겠습니다.
equltable만 Extension파일로 구분하여 추가하는 형태로 변경했습니다.

havilog replied:
Sendable은 비동기 컨텍스트에 사용할 일이 있을 때 사용하는데, 앵간하면 준수해서 나쁠 것은 없다고 생각해요
Identifiable은 리스트에 넣는 아이템 같이 아이디가 반드시 필요한 경우 필수적으로 구현해주어야합니다

havilog replied:
Extension 파일에서
static func == 을 구현하지 않아도 될 거 같은데 굳이 파일을 분리해서 구현해주신 이유는 뭘까요?
가장 하위 모델은 이미 프로퍼티들이 equltable을 만족하고 있어서 모델에 바로 붙여주는게 더 가독성이 좋을 것 같아요

havilog commented on BoxOfficeTests/DecodingTests.swift:
```diff
@@ -0,0 +1,142 @@
+final class DecodingTests: XCTestCase {
+    let decoder = JSONDecoder()
+
+    func test_init으로_만든_BoxOfficeResult와_JSON을_디코딩한_결과가_같은지_확인() throws {
```
요건 개인적인 생각이고 사람마다 생각이 다를 수도 있는 부분인데요,

테스터블 한 코드가 좋은 코드다
는 대부분 동의할텐데

테스트 코드가 꼭 필요하다
는 호불호가 갈리는 부분이에요

저는 비지니스로직에는 있으면 좋다
라는 생각을 가지고 있는데요

UI나 단순 액션들의 경우는 오히려 생산성 저하를 가져올 수 있어요

지금 이 테스트 코드의 경우는 모델을 잘 구현했는지 보다는 swift의 decoder를 테스트하는 느낌이 들어요

따라서 테스트는 뷰에서 어떤 액션이 일어났을 때
원하는 결과값이 도출되었는지
사용하지 않는 의존성이 사용되지는 않았는지
등을 테스트하는게 좋다고 생각합니다

Hi-sop replied:
모델을 잘 구현했는지 확인할 다른 방법이 떠오르지 않습니다...
다음스텝에 테스트더블과 함께 네트워크 대신 test.json으로 바꿔치기하는 방식정도만 생각납니다
괜찮은 방법 있을까요?

말씀하신 부분 중 'UI나 단순 액션들의 경우는 오히려 생산성 저하를 가져올 수 있다' 에 JSON모델 테스트는 속하지 않는거겠죠?

havilog replied:
제 기준에서 JSON모델 테스트는.. 테스트 코드로 테스트 할 만한 아이는 아니라고 생각해요!
실제 서버에서 오는 응답이 달라질 경우에는 어쩌피 터지는거라

다만 요구사항에 있으니 구현하셔야 해서 상충되는 부분은 있지만,
야곰에서는 이렇게 생각하고, 저는 이렇게 생각한다 정도로 이해해주시면 될 거 같아요

havilog commented on .DS_Store:
dsstore같은 파일들은 깃이그노어로 안올리는게 좋아보여요!

Hi-sop replied:
깃 이그노어에 추가되어있는것으로 보이는데 같이 올라갔네요.. 다시 확인해보겠습니다

havilog reviewed (approved):
수고하셨어요 히섭~!
오탈자 방지에 대한 꿀팁은 ... 매의 눈.. 정도?
제가 아는 건 딱히 없네요 ㅎㅎ.. ㅠㅠ
수정하시고 싶으신 부분 있으면 수정하시고 알려주시면 머지할게요~

Hi-sop committed:
refactor: MovieStatistics -> DailyBoxOffice 네이밍 변경

Hi-sop committed:
refactor: 타입에 private 접근제어 추가

Hi-sop committed:
refactor: Extension파일 생성,  타입의 private접근제어 삭제

BoxOfficeResult파일이름 변경
Extension으로 타입에 접근시 private으로 인한 컴파일 에러 발생 -> private 다시 제거

Hi-sop committed:
CodingKeys private 추가

Hi-sop committed:
refactor: DailyBoxOffice Extension삭제
